packages/@aws-cdk/aws-iam/lib/private/merge-statements.ts

```diff
@@ -0,0 +1,241 @@
+// IAM Statement merging
+//
+// See docs/policy-merging.als for a formal model of the logic
+// implemented here.
+
+const LENGTH_CACHE_SYM = Symbol('jsonLength');
+
+/**
+ * Merge as many statements as possible to shrink the total policy doc, modifying the input array in place
+ *
+ * If there are multiple merges to be made (which may involve the same
```

Comment:
From the description in `PolicyDocumentProps`, it sounds like the merge is associative, commutative and idempotent (but I may be missing something). If that's the case, the order shouldn't matter. And if it's not the case, can we make it so?

Reply:
That's a good question. I can ask the Alloy model whether that is true as well.

Reply:
It is! Obvious in hindsight, as always 😆. Thanks for pointing this out. Combined with doing subset analysis (instead of `==` analysis), I think this is now much simpler and more optimal.

Had to move some code around for statement normalization.

Reply:
Well nevermind. That subset-based merging I said we could do is in fact NOT safe, and leads to the exact same error as the one we had last time around.

Reply:
But merging is still associative.

Reply:
What's the counterexample for commutativity? My intuition is that we can define a merge operation that forms a semilattice, which induces a partial order (if not the subset relation you tried, maybe some variation of it). Then, navigate the DAG and keep only the statements that are "leaves" in the graph (i.e., not a predecessor of any other statement).